From b09b22037467b82b73edac5b998bb4d0d9dbf9b9 Mon Sep 17 00:00:00 2001
From: Marian Chereji <marian.chereji@freescale.com>
Date: Mon, 29 Oct 2012 21:36:24 +0000
Subject: [PATCH 347/518] dpa_classifier: Fix 64bit kernel/32bit user
 applications compatibility

The compatibility mode needed some adjustments to get synchonized
with the latest dpa_classifier API updates.

Signed-off-by: Marian Chereji <marian.chereji@freescale.com>
[Grabbed from the branch, LINUX_IR5.2.0, of
https://git.freescale.com/git-private/cgit.cgi/ppc/alu-b4860/linux.git.]
Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
---
 .../staging/fsl_dpa_offload/dpa_classifier_ioctl.h |    3 ++-
 .../staging/fsl_dpa_offload/wrp_dpa_classifier.c   |    4 ++--
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/fsl_dpa_offload/dpa_classifier_ioctl.h b/drivers/staging/fsl_dpa_offload/dpa_classifier_ioctl.h
index 130f68d7..ffa8a88 100644
--- a/drivers/staging/fsl_dpa_offload/dpa_classifier_ioctl.h
+++ b/drivers/staging/fsl_dpa_offload/dpa_classifier_ioctl.h
@@ -185,6 +185,7 @@ struct dpa_cls_compat_tbl_params {
 		struct dpa_cls_tbl_indexed_params	indexed_params;
 		struct dpa_cls_tbl_exact_match_params	exact_match_params;
 	};
+	unsigned int			prefilled_entries;
 };
 
 struct compat_ioc_dpa_cls_tbl_params {
@@ -346,7 +347,7 @@ struct dpa_cls_compat_hm_nat_params {
 	union {
 		struct dpa_cls_hm_traditional_nat_params	nat;
 		struct dpa_cls_hm_nat_pt_params			nat_pt;
-	} params;
+	};
 	uint16_t	sport;
 	uint16_t	dport;
 	compat_uptr_t	fm_pcd;
diff --git a/drivers/staging/fsl_dpa_offload/wrp_dpa_classifier.c b/drivers/staging/fsl_dpa_offload/wrp_dpa_classifier.c
index 7b047cb..f47a00d 100644
--- a/drivers/staging/fsl_dpa_offload/wrp_dpa_classifier.c
+++ b/drivers/staging/fsl_dpa_offload/wrp_dpa_classifier.c
@@ -2286,8 +2286,8 @@ int dpa_cls_hm_nat_params_compatcpy(
 	kparam->nat_params.flags = uparam->nat_params.flags;
 	kparam->nat_params.proto = uparam->nat_params.proto;
 	kparam->nat_params.type = uparam->nat_params.type;
-	memcpy(&kparam->nat_params.params, &uparam->nat_params.params,
-		sizeof(kparam->nat_params.params));
+	memcpy(&kparam->nat_params.nat_pt, &uparam->nat_params.nat_pt,
+		sizeof(kparam->nat_params.nat_pt));
 
 	kparam->nat_params.fm_pcd = compat_ptr(uparam->nat_params.fm_pcd);
 	kparam->next_hmd = uparam->next_hmd;
-- 
1.7.5.4

